[SQL] Program with MAX applied to ROW type fails to compile

For a SQL program such as:

```
CREATE TABLE row_tbl(
c1 INT,
c2 VARCHAR,
c3 VARCHAR);

CREATE MATERIALIZED VIEW v_max AS SELECT
MAX(ROW(c1, c2, c3)) AS c1
FROM row_tbl;
```

The compilation fails with a Rust error:
**Rust error: the Rust code generated based on the SQL (possibly combined with any user-provided UDF Rust and TOML code) failed to compile.**

With the following logs:
```
stderr:
   Compiling project-9a0124b0dee4c6f8308393f1674e239dbb718e45b1ce891fbc42338408c659cd v0.1.0 (/Users/rivuadhikari/.feldera/compiler/rust-compilation/projects/project-9a0124b0dee4c6f8308393f1674e239dbb718e45b1ce891fbc42338408c659cd)
error[E0308]: mismatched types
   --> projects/project-9a0124b0dee4c6f8308393f1674e239dbb718e45b1ce891fbc42338408c659cd/src/main.rs:164:67
    |
164 |                 (*t_10) = Tup1::new(agg_max_N_((*t_10).0.clone(), (*t_3).0.clone()))
    |                                     ----------                    ^^^^^^^^^^^^^^^^ expected `Tup3<Option<i32>, Option<...>, ...>`, found `Tup3<i32, Option<String>, ...>`
    |                                     |
    |                                     arguments to this function are incorrect
    |
    = note: expected struct `dbsp::utils::Tup3<std::option::Option<i32>, _, _>`
               found struct `dbsp::utils::Tup3<i32, _, _>`
help: the return type of this call is `dbsp::utils::Tup3<i32, std::option::Option<std::string::String>, std::option::Option<std::string::String>>` due to the type of the argument passed
   --> projects/project-9a0124b0dee4c6f8308393f1674e239dbb718e45b1ce891fbc42338408c659cd/src/main.rs:164:37
    |
164 |                 (*t_10) = Tup1::new(agg_max_N_((*t_10).0.clone(), (*t_3).0.clone()))
    |                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^----------------^
    |                                                                   |
    |                                                                   this argument influences the return type of `agg_max_N_`
note: function defined here
   --> /Users/rivuadhikari/feldera/crates/sqllib/src/aggregates.rs:307:1
    |
307 | universal_aggregate!(agg_max);
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    = note: this error originates in the macro `universal_aggregate` (in Nightly builds, run with -Z macro-backtrace for more info)
```

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/backend/rust/ToRustInnerVisitor.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/backend/rust/ToRustInnerVisitor.java
@@ -794,11 +794,25 @@ public class ToRustInnerVisitor extends InnerVisitor {
 
         if (sourceType.is(DBSPTypeTuple.class)) {
             // should have been eliminated
-            this.unimplementedCast(expression);
+            if (!this.compact)
+                this.unimplementedCast(expression);
+            // we are dumping DOT
+            this.builder.append("(");
+            expression.type.accept(this);
+            this.builder.append(")");
+            expression.source.accept(this);
+            return VisitDecision.STOP;
         }
         if (destType.is(DBSPTypeTuple.class)) {
             // should have been eliminated
-            this.unimplementedCast(expression);
+            if (!this.compact)
+                this.unimplementedCast(expression);
+            // we are dumping DOT
+            this.builder.append("(");
+            expression.type.accept(this);
+            this.builder.append(")");
+            expression.source.accept(this);
+            return VisitDecision.STOP;
         }
 
         functionName = "cast_to_" + destType.baseTypeWithSuffix() +

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/AggregateCompiler.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/AggregateCompiler.java
@@ -63,6 +63,7 @@ import org.dbsp.sqlCompiler.ir.expression.literal.DBSPVecLiteral;
 import org.dbsp.sqlCompiler.ir.type.DBSPType;
 import org.dbsp.sqlCompiler.ir.type.derived.DBSPTypeTuple;
 import org.dbsp.sqlCompiler.ir.type.IsNumericType;
+import org.dbsp.sqlCompiler.ir.type.primitive.DBSPTypeBaseType;
 import org.dbsp.sqlCompiler.ir.type.primitive.DBSPTypeBinary;
 import org.dbsp.sqlCompiler.ir.type.primitive.DBSPTypeBool;
 import org.dbsp.sqlCompiler.ir.type.primitive.DBSPTypeDouble;
@@ -385,9 +386,21 @@ public class AggregateCompiler implements ICompilerComponent {
     public DBSPExpression aggregateOperation(
             CalciteObject node, DBSPOpcode op,
             DBSPType type, DBSPExpression left, DBSPExpression right, @Nullable DBSPExpression filter) {
+        assert op.isAggregate;
         DBSPType leftType = left.getType();
         DBSPType rightType = right.getType();
         DBSPType resultType = type.withMayBeNull(leftType.mayBeNull || rightType.mayBeNull);
+        assert type.sameType(leftType);
+        // For all types the nullability of leftType and rightType may differ.
+        // For recursive types the element types of leftType and rightType may be wrong from Calcite,
+        // because Calcite sometimes assumes (but not always!) that field types are always nullable.
+        if (!leftType.withMayBeNull(rightType.mayBeNull).sameType(rightType)) {
+            if (!rightType.is(DBSPTypeBaseType.class)) {
+                // These can also be different DECIMAL types
+                right = right.cast(leftType.withMayBeNull(rightType.mayBeNull));
+            }
+        }
+
         DBSPExpression binOp = new DBSPConditionalAggregateExpression(
                 node, op, resultType, left.applyCloneIfNeeded(), right.applyCloneIfNeeded(), filter);
         return binOp.cast(type);

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/inner/ExpandCasts.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/inner/ExpandCasts.java
@@ -96,7 +96,6 @@ public class ExpandCasts extends InnerRewriteVisitor {
     }
 
     DBSPExpression convertToStruct(DBSPExpression source, DBSPTypeTuple type) {
-        // Convert a tuple to a VARIANT MAP indexed by the field names
         List<DBSPExpression> fields = new ArrayList<>();
         assert type.originalStruct != null;
         DBSPType sourceType = source.getType();

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/inner/Simplify.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/inner/Simplify.java
@@ -126,7 +126,7 @@ public class Simplify extends InnerRewriteVisitor {
         this.push(expression);
         DBSPExpression source = this.transform(expression.expression);
         this.pop(expression);
-        DBSPExpression result = source.applyClone();
+        DBSPExpression result = source.applyCloneIfNeeded();
         if (source.is(DBSPCloneExpression.class)) {
             result = source;
         }

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/ir/expression/DBSPExpression.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/ir/expression/DBSPExpression.java
@@ -192,6 +192,9 @@ public abstract class DBSPExpression
     public DBSPExpression applyCloneIfNeeded() {
         if (this.getType().hasCopy())
             return this;
+        if (this.is(DBSPBaseTupleExpression.class))
+            // No need to clone a Tuple constructor
+            return this;
         return this.applyClone();
     }
 

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/IncrementalRegressionTests.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/IncrementalRegressionTests.java
@@ -50,6 +50,23 @@ public class IncrementalRegressionTests extends SqlIoTest {
     }
 
     @Test
+    public void issue3145() {
+        this.compileRustTestCase("""
+                CREATE TABLE array_tbl(c1 INT ARRAY NOT NULL);
+                CREATE MATERIALIZED VIEW max_emp AS SELECT
+                MAX(c1) AS max FROM array_tbl WHERE FALSE;""");
+    }
+
+    @Test
+    public void issue3119() {
+        this.compileRustTestCase("""
+                CREATE TABLE row_tbl(id INT, c1 INT NOT NULL, c2 VARCHAR, c3 VARCHAR);
+                CREATE MATERIALIZED VIEW v_max AS SELECT
+                MAX(ROW(c1, c2, c3)) AS c1
+                FROM row_tbl;""");
+    }
+
+    @Test
     public void wrongLateness() {
         this.statementsFailingInCompilation("""
                 create table t (

